### DISCUSSION TITLE
M220/M221 Open discussion

### DISCUSSION THREAD

**Discussion by J0hanss0n** [bigtreetech/BIGTREETECH-TouchScreenFirmware#2338] [General]
@kisslorand: I have a new information:
I examined Marlin code and there's no DMA involved in Serial communication for STM32. Just single byte interrupts. (The comment mentioning DMA you found seems related to SDIO.) So it's even more obvious that answering of M220 can happen during reception of M221.
But now I have no new theory, why a corruption occurs.
Regards, Johannes

**Comment by kisslorand** [comment]
It seems very complicated. 
@digant suggested to use only one serial port in Marlin or to check only one serial for incoming commands. I edited Marlin to check incoming commands only from TFT and for an hour or so there's no error. I know it doesn't mean anything because the error sometimes doesn't show up for hours.


>  So it's even more obvious that answering of M220 can happen during reception of M221. But now I have no new theory, why a corruption occurs. 

So, if no DMA is involved than why an answer from Marlin leads to skipped bytes?
Another one, why is mustStoreScript() implemented? I do not think this one is a new issue, that one wasn't implemented for no reason.

**Comment by kisslorand** [comment]
A few hours later:  there was a nice "No SD Card" on my TFT. So having just one serial enabled in Marlin hasn't fixed  loosing bytes from TFT.
Actually based on the "Notification" log of the TFT (click the upper part of the touchscreen) there were more "No SD Card" messages received during this period.

**Comment by J0hanss0n** [comment]
Just a follow-up: I'm still checking Marlin for a reason for the lost bytes. Already down in Arduino core for STM32, but no theory yet.

**Comment by kisslorand** [comment]
My head is dizzy checking Marlin... I haven't find anything yet too... I have a strong feeling it is MCU related.

**Comment by J0hanss0n** [comment]
@digant73 @kisslorand 
I had not much time to do more work on this topic but one thing puzzles me, perhaps one of you can answer:
From my experience as embedded developer I'd never implement a serial communication without a protocol featuring a checksum.
I had now a look at Marlin and at BigtreeTFT firmware, and surprise, both can use line numbering and checksum calculation. So, how can an error like the observed one occur, because a missing byte in the data stream (wherever it may get lost), should lead to an invalid checksum.
My conclusions so far:
Checksum and Line numbering is not used when infoMachineSettings.firmwareType != FW_REPRAPFW (Function sendCmd of interfaceCmd.c) and I assume the firmwareType is set to FW_MARLINFW in parseAck.c, which means no Line numbering and no checksum.
Could you please explain or confirm my findings?

Johannes

**Comment by kisslorand** [comment]
I confirm that in case of Marlin there's no checksum nor line numbering used in UART communication.
The reasoning must be that it is unlikely that a communication error might occur in such a short distance between the TFT and the mainboard if everything is done properly. On the other hand on a 8bit motherboard it would surely affect printing if line numbering, checksum check would be implemented. Just think about the issue when you need to print a curve with a huge number of tiny segments.
I haven't had much time neither to check Marlin further and I am still very surprised that on the Marlin's Github repository no one is mentioning this issue.

**Reply by J0hanss0n** [reply]
>...if everything is done properly.

You name it.
I'm not sure, but I'd think even on a 8bit mainboard the computing overhead (xor and two strtol) would be acceptable. But nevertheless the additional line numbers and checksums consume bandwidth on the serial line.
I have the strong feeling, that the missing bytes are caused by blocked interrupts (more than 100us @115200baud) on Marlin side. But there is also information that the error occurs only with newer TFT firmware versions keeping the same Marlin version.
Strange.

**Comment by kisslorand** [comment]
It's more complicated than this.
The error occurs on newer TFT versions because it is appearing only with very new mainboards, so most probably people upgraded their FW.
I remember a discussion in Marlin where somebody changed what timer to handle the servo pin on SKR PRO (same MCU as on SKR v2) and it was a warning that changing that can have an unknown impact in the future. The level of discussion was to high for my knowledge to be able to reproduce the concerns discussed there.

I forgot to mention (again) that I have done intensive HW log between the TFT and the motherboard. There were many times the TFT sent M221 and Marlin responded "No SD card" so it clearly interpreted M221 as M21. The monitoring was done by sniffing the very wires between TFT and motherboard.

**Comment by J0hanss0n** [comment]
I have always your measurements in mind. That's the reason I'm seeing the problem on Marlin side (blocked Interrupts).
On a side note: In your recent PR you are referring a baudrate of 2500000 and 40us for 10 bytes. Are you really running the UART @ 2.5Mbit/s? With 250000 (0.25Mbit/s) 10 bytes would be 400us on the wire.

Edit:
https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/issues/2295#issuecomment-1000183653
There's a comment mentioning that the TFT firmware version is also relevant.

**Reply by kisslorand** [reply]
>On a side note: In your recent PR you are referring a baudrate of 2500000 and 40us for 10 bytes. Are you really running the UART @ 2.5Mbit/s? With 250000 (0.25Mbit/s) 10 bytes would be 400us on the wire.

It was a typo, I am running at 250,000 and it's indeed 400us. I was thinking baud/s and mixed up things.
Thanks for the observation, I corrected the info in my PR.

>#2295 (comment)
There's a comment mentioning that the TFT firmware version is also relevant.

I remember at some point BTT switched the libraries used at the point when STM released version 8 or 9.  I do not remember if the released version still uses the STM libraries or not.

**Comment by kisslorand** [comment]
I might have found something. Testing it...
